C2PA-305: More detailed font errors

brogdonm opened the pull request:
## Changes in this pull request

An extension to the previous PR which now converts all internal font_io/sft_io/woff_io calls to use `FontError` as the error type. And then all of the crate level trait implementations, the error is converted to a `crate::error::Error`.

I also took the time to delete unused errors and split errors that were being used generically into more detailed errors.

For the `SaveError` I spun out a separate error enum to scope the errors for more details on the save errors.

## Checklist
- [X] This PR represents a single feature, fix, or change.
- [X] All applicable changes have been documented.

brogdonm committed:
Make sure all calls within font_io use the FontError types.

brogdonm committed:
Switch over the sft_io to use all FontError for internal calls, mapping to crate error on trait impls.

brogdonm committed:
Fix formatting issue.

brogdonm committed:
Fix up woff_io to use FontError as well for internal calls.

brogdonm committed:
Get rid of generic LoadError and replace with other errors.

brogdonm committed:
Split the BadMissing head table error into 2 separate errors.

brogdonm committed:
Expand the save error to be more specific.

brogdonm committed:
Remove the old generic SaveError and update for woff as well.

brogdonm committed:
Clean up the use of Result

chriskushmonotype commented on sdk/src/asset_handlers/font_io.rs:
```diff
@@ -16,26 +16,42 @@ use std::io::{Read, Seek, SeekFrom, Write};
+    #[error("Invalid named table: {0}")]
+    InvalidNamedTable(&'static str),
+
+    //// The JUMBF data was not found.
```
🔢

vancejc-mt commented on sdk/src/asset_handlers/sfnt_io.rs:
```diff
@@ -891,15 +903,15 @@ where
     // Read the font from the input stream
-    let mut font = SfntFont::from_reader(input_stream).map_err(|_| FontError::LoadError)?;
+    let mut font =
+        SfntFont::from_reader(input_stream).map_err(|_| FontError::DeserializationError)?;
```
Is this mapping necessary?  They both throw the same error type right?

brogdonm replied:
Yeah, I was curious about this too. The previous code was basically remapping the underlying error types into a `DeserializationError`. I wonder if we do the same thing here as the `FontSaveError`, have a `FontLoadError` with the other details in it to capture that the fact it failed to load with more details why it failed to load. What do you think?

vancejc-mt replied:
Uh.. I don't know if it's necessary; the `from_reader()` should give you all the details of why it's failing - adding more information here would just give.. context about why you were reading? Might not be all that helpful

brogdonm replied:
Ok, thanks. This will just remove the `DeserializeError` then, updating...

vancejc-mt commented on sdk/src/asset_handlers/sfnt_io.rs:
```diff
@@ -1076,14 +1088,15 @@ where
 fn read_c2pa_from_stream<T: Read + Seek + ?Sized>(reader: &mut T) -> Result<TableC2PA> {
-    let sfnt = SfntFont::from_reader(reader).map_err(|_| FontError::LoadError)?;
+    // Convert all errors from the reader to a deserialization error.
+    let sfnt = SfntFont::from_reader(reader).map_err(|_| FontError::DeserializationError)?;
```
Same here - I don't think the mapping is necessary?

vancejc-mt commented on sdk/src/asset_handlers/sfnt_io.rs:
```diff
@@ -1129,23 +1145,27 @@ impl CAIWriter for SfntIO {
         store_bytes: &[u8],
-    ) -> Result<()> {
-        add_c2pa_to_stream(input_stream, output_stream, store_bytes)
+    ) -> crate::error::Result<()> {
```
So I think I prefer the `.map_err(wrap_font_err)` way of mapping the error here, the `Ok()` seems a little too tricky

vancejc-mt commented on sdk/src/asset_handlers/woff_io.rs:
```diff
@@ -844,15 +844,15 @@ where
     // Read the font from the input stream
-    let mut font = WoffFont::from_reader(input_stream).map_err(|_| FontError::LoadError)?;
+    let mut font =
+        WoffFont::from_reader(input_stream).map_err(|_| FontError::DeserializationError)?;
```
Unnecessary `map_err` here?

vancejc-mt commented on sdk/src/asset_handlers/woff_io.rs:
```diff
@@ -1071,18 +1071,16 @@ where
-    let c2pa_table: Option<TableC2PA> = match woff.tables.get(&C2PA_TABLE_TAG) {
-        None => None,
+    // Convert all errors from the reader to a deserialization error.
+    let woff = WoffFont::from_reader(reader).map_err(|_| FontError::DeserializationError)?;
```
Unneccessary `map_err` here as well?

vancejc-mt commented on sdk/src/asset_handlers/sfnt_io.rs:
```diff
@@ -322,7 +322,9 @@ impl SfntFont {
                     // We added some other number of tables
-                    return Err(wrap_font_err(FontError::SaveError));
+                    return Err(FontError::SaveError(
+                        FontSaveError::UnexpectedNumberOfTables,
```
I think here instead maybe we have a `TooManyTablesAdded` error to be more specific

vancejc-mt commented on sdk/src/asset_handlers/sfnt_io.rs:
```diff
@@ -332,13 +334,17 @@ impl SfntFont {
                     // We added some other number of tables. Weird, right?
-                    return Err(wrap_font_err(FontError::SaveError));
+                    return Err(FontError::SaveError(
+                        FontSaveError::UnexpectedNumberOfTables,
```
I think here instead maybe we have a `TooManyTablesRemoved` error to be more specific

vancejc-mt commented on sdk/src/asset_handlers/woff_io.rs:
```diff
@@ -1127,23 +1128,27 @@ impl CAIWriter for WoffIO {
-    ) -> Result<Vec<HashObjectPositions>> {
-        get_object_locations_from_stream(self, input_stream)
+    ) -> crate::error::Result<Vec<HashObjectPositions>> {
+        Ok(get_object_locations_from_stream(self, input_stream)?)
```
Same stuff here and following, with `Ok()` vs `map_err(wrap_font_err)`

vancejc-mt reviewed (changes requested):
Slight changes, a few unnecessary maps maybe?

brogdonm committed:
PR review part 1

brogdonm committed:
PR review part 2.

brogdonm committed:
PR review part 3.

Split the UnexpectedNumberOfTables into two new errors with exactly the issue reported.

brogdonm committed:
Grab fmt changes.

brogdonm committed:
include the IO error in the message.

vancejc-mt reviewed (approved):
Changes look good to me ✅ builds and runs like a champ
